src/client/eppo-precomputed-client.spec.ts

```diff
@@ -0,0 +1,706 @@
+  });
+
+  it('returns default value when key does not exist', async () => {
+    const client = new EppoPrecomputedClient(storage);
+    const nonExistentFlag = 'non-existent-flag';
+    expect(client.getStringAssignment(nonExistentFlag, 'default')).toBe('default');
+  });
+
+  it('logs variation assignment with correct metadata', () => {
+    const mockLogger = td.object<IAssignmentLogger>();
+    storage.setEntries({ [precomputedFlagKey]: mockPrecomputedFlag });
+    const client = new EppoPrecomputedClient(storage);
+    client.setAssignmentLogger(mockLogger);
+
+    client.getStringAssignment(precomputedFlagKey, 'default');
+
+    expect(td.explain(mockLogger.logAssignment).callCount).toEqual(1);
+    const loggedEvent = td.explain(mockLogger.logAssignment).calls[0].args[0];
+
+    expect(loggedEvent.featureFlag).toEqual(precomputedFlagKey);
+    expect(loggedEvent.variation).toEqual(mockPrecomputedFlag.variationKey);
+    expect(loggedEvent.allocation).toEqual(mockPrecomputedFlag.allocationKey);
+    expect(loggedEvent.experiment).toEqual(
+      `${precomputedFlagKey}-${mockPrecomputedFlag.allocationKey}`,
+    );
+  });
+
+  it('handles logging exception', () => {
+    const mockLogger = td.object<IAssignmentLogger>();
+    td.when(mockLogger.logAssignment(td.matchers.anything())).thenThrow(new Error('logging error'));
+
+    storage.setEntries({ [precomputedFlagKey]: mockPrecomputedFlag });
+    const client = new EppoPrecomputedClient(storage);
+    client.setAssignmentLogger(mockLogger);
+
+    const assignment = client.getStringAssignment(precomputedFlagKey, 'default');
+
+    expect(assignment).toEqual('variation-a');
+  });
+
+  describe('assignment logging deduplication', () => {
+    let client: EppoPrecomputedClient;
+    let mockLogger: IAssignmentLogger;
+
+    beforeEach(() => {
+      mockLogger = td.object<IAssignmentLogger>();
+      storage.setEntries({ [precomputedFlagKey]: mockPrecomputedFlag });
+      client = new EppoPrecomputedClient(storage);
+      client.setAssignmentLogger(mockLogger);
+    });
+
+    it('logs duplicate assignments without an assignment cache', async () => {
+      client.disableAssignmentCache();
+
+      client.getStringAssignment(precomputedFlagKey, 'default');
+      client.getStringAssignment(precomputedFlagKey, 'default');
+
+      // call count should be 2 because there is no cache.
+      expect(td.explain(mockLogger.logAssignment).callCount).toEqual(2);
+    });
+
+    it('does not log duplicate assignments', async () => {
+      client.useNonExpiringInMemoryAssignmentCache();
+      client.getStringAssignment(precomputedFlagKey, 'default');
+      client.getStringAssignment(precomputedFlagKey, 'default');
+      // call count should be 1 because the second call is a cache hit and not logged.
+      expect(td.explain(mockLogger.logAssignment).callCount).toEqual(1);
+    });
+
+    it('logs assignment again after the lru cache is full', async () => {
```

Comment:
I'm glad you are thinking about testing the assignment logging and deduplication - as you mentioned there is some duplication here. Could you consider a way in the future to have the two eppo clients share these?